lib/cyrusdb_twoskip.c

```diff
@@ -1427,16 +1436,28 @@ static int opendb(const char *fname, int flags, struct dbengine **ret, struct tx
         if (r) goto done;
     }
 
+    if (mytid) {
+        int shared = flags & CYRUSDB_SHARED;
+        int iswlocked = mappedfile_iswritelocked(db->mf);
+        if ((iswlocked && !shared) || (!iswlocked && shared)) {
+            // already have correct lock!
```

Comment:
Not sure I trust this assertion, what if it's unlocked and needs to be read locked?

Ahh, looks like earlier we either got a read lock, or a write lock, or errored out.  So the `iswlocked` boolean represents write vs read, not write vs read-or-unlocked.

Reply:
Yep, as you noted.

We could also be using our own variable set in the locking code higher up, but it's cheap to just ask the mappedfile object.